readme edits (#6)
piiano/cellotape: README.md

> ![Cellotape mascot](./Cellotape-Gopher.png)

Reviewer: The file name is still `cellotape-gopher.png` in this PR.
Notice that case sensitive only file name changes are ignored by regular `git add .`.
You can use this command to tell git the file is reanmed:
`git mv -f cellotape-gopher.png Cellotape-Gopher.png`

Author: Sorry, overzealous correction of capitalization on "go". Reverting this.

> ![Cellotape mascot](./Cellotape-gopher.png)

Reviewer: ```suggestion
![Cellotape mascot](./cellotape-gopher.png)
```

> **🚧 Cellotape is in Beta 🚧**
> 
> Please note that this is a beta version, and the API may change.

A type-safe approach to HTTP routing with OpenAPI in Go. The project aims to 
simplify the development of REST APIs with OpenAPI by enabling a 
**design-first** approach.

Cellotape loads an OpenAPI spec and uses it as a router to call relevant handlers 
in your code. The handler signatures are validated with your OpenAPI spec to verify
 your code is implementing the design correctly.

- [Concepts](#concepts)
  - [Included features](#included-features)
  - [What this project isn't doing](#what-this-project-isnt-doing)
- [Get started](#get-started)
- [Loading an OpenAPI spec (`router.OpenAPISpec`)](#loading-openapi-spec-routeropenapispec)
  - [Using FS embedding (recommended)](#using-fs-embedding-recommended)
  - [Read file at runtime](#read-file-at-runtime)

>   - [Define programmatically in Go](#define-programmatically-in-Go)

Reviewer: ```suggestion
  - [Define programmatically in Go](#define-programmatically-in-go)
```

Author: As above - reverted

> - [Initialize new HTTP router (`router.OpenAPIRouter`)](#init-new-http-router-routeropenapirouter)

Reviewer: ```suggestion
- [Initialize new HTTP router (`router.OpenAPIRouter`)](#initialize-new-http-router-routeropenapirouter)
```

Author: Done

The Go ecosystem provides many packages and frameworks for building HTTP servers 
for REST APIs. Most of them rely on concepts from the built-in 
[net/http](https://pkg.Go.dev/net/http) package to define HTTP request handlers.

The issue Cellotape solves is the lack of type information when working with such 
handlers.

Instead of having the same untyped signature for all handlers, Cellotape handlers 
use generics to define for each handler the types of its body, path parameters, 
query parameters, and responses.

This extra type information enables Cellotape to define handlers that are 
validated at runtime with an OpenAPI specification.

### Included features

- Load, parse, and validate an OpenAPI spec (`router.OpenAPISpec`).
- Initialize an HTTP router driven from an OpenAPI spec (`router.OpenAPIRouter`).
- An SDK that defines strongly typed handlers mapped to spec operations.
- Verify handler signature compatibility with spec operations.
  During initialization, this enforces that the handlers correctly implement the 
  spec. The validated components include:
  - Request body schema
  - Path parameters
  - Query parameters
  - Responses

> - Support for middleware chain and group mechanisms that enable middleware to 
>   apply specific operations or specific groups.

Reviewer: ```suggestion
- Support for middleware chain and group mechanisms that enable middleware to 
  be applied for specific operations or specific groups.
```

Author: Incorporated

- Compatibility with the `HTTP.Handler` interface for the router and middleware 
  to enable easy integration of the router into any popular framework.
- Support for custom content types to align with content types defined in the spec.
  This can be done by implementing the `router.ContentType` interface.
- Support for customization of validation behavior and other configuration using
  `router.Options`.

>   See the documentation on the [`router.Options`](./router/options.Go) struct 

Reviewer: ```suggestion
  See the documentation on the [`router.Options`](./router/options.go) struct 
```

Author: Reverted

  for details of the options.

### What this project isn't doing

- It's not a code generator.
  Code generators have their issues, such as minimal control over the generated 
  code and the overwriting of manual edits whenever the spec changes.

> - It's not using comments on your code. Some approaches might suggest you 
>   comment code in a way that produces an OpenAPI spec that can be compared to the 
>   spec.
>   The issue with this approach is that the spec, code, and comments can get out 
>   of sync too easily. Primarily when HTTP handlers in Go provide an opaque 
>   interface with no type information.

Reviewer: I think sentence is still a bit unclear (it wasn't so clear also before).

```suggestion
- It's not using comments on your code. Some approaches might suggest you 
  comment code in a way that produces an OpenAPI spec (a code first approach).
  
  Then, in order to use a design first approach the produced spec can be compared to the 
  desired spec.
  
  The issue with this approach is that the spec, code, and comments can get out 
  of sync too easily. Primarily when HTTP handlers in Go provide an opaque 
  interface with no type information.
```

Author: Suggested a further improvement in the update

Cellotape use of strongly typed handlers can help create an implementation for 
your API that never gets out of sync and is easily maintained.

## Get started

Add Cellotape to your project using `go get`:

```bash

> go get github.com/piiano/Cellotape@latest

Reviewer: ```suggestion
go get github.com/piiano/cellotape@latest
```

Author: Reverted

}
```

This code uses `router.NewSpecFromData` to load the OpenAPI spec.

> You initialize the Cellotape router with `router.NewOpenAPIRouter`. 

Reviewer: ```suggestion
You initialize the Cellotape router with `router.NewOpenAPIRouter`.
```

Author: Done

> server again.
> 
> Cellotape reports any incompatibility, enabling you to check that your 
> implementation and spec are in sync. 

## Loading OpenAPI spec (`router.OpenAPISpec`)

To initialize a new Cellotape HTTP router you must first load an OpenAPI spec. 

The OpenAPI spec defines URL paths and HTTP methods for various operations. 
Cellotape routes HTTP calls on these paths and methods for the relevant handler 
implementation of each operation. Cellotape also validates that the handler 
correctly implements the request and response defined in the spec.

There are several ways you can load the OpenAPI spec for use with Cellotape.

### Using FS embedding (recommended)

Using [Go embedding](https://pkg.go.dev/embed), you embed the OpenAPI YAML or 
JSON file with the compiled binary of your app. For example, embedding an 
`openapi.yaml` like this:

```go
//go:embed openapi.yaml
var specData []byte
```

This enables you to initialize the spec from its bytes with `router.NewSpecFromData` 
like this:

```go
spec, err := router.NewSpecFromData(specData)
```

`router.NewSpecFromData` returns the loaded spec object or an error if the object 

> fails to parse and validate.

Reviewer: I used **and** because Go functions can return multiple values.
In this case:
```go
spec, err := router.NewSpecFromData(specData)
```
The function returns `spec` **and** an `err` (error).
The error is `nil` (empty) if the spec parsed and validated correctly but it is still a `nil` value returned from the function for the err value.

Please let me know if you still think it should be **or**.
Notice this is a very common pattern in Go.

Author: Thanks, understood, reverted

Reviewer: Technically it returns both at the same time but with one of them having a "zero value" or `nil`.
Few examples from [effective go](https://go.dev/doc/effective_go):
- https://go.dev/doc/effective_go#:~:text=return%20a%20count%20and%20an%20error
- https://go.dev/doc/effective_go#:~:text=The%20method%20returns%20the%20number%20of%20bytes%20read%20and%20an%20error%20value
- https://go.dev/doc/effective_go#:~:text=returns%20a%20value%20and%20an%20error

})
```

Notice how the multiple fields of the responses struct enable you to send the 
response you want based on your implementation logic in a way that covers all the 
responses defined the spec.

Each response type is validated for compatibility with the schema defined in each
 response of the spec `responses`.

## Examples

Examples that show how to use Cellotape.

### Hello World API example

This example uses an [openapi.yaml](./examples/hello_world_example/openapi.yaml)
to initialize the server and map to the relevant handlers. Visit the 
[Hello World Example](./examples/hello_world_example) to see how it works.

### TODO List API example

This example use an [openapi.yaml](./examples/todo_list_app_example/openapi.yaml)
to initialize the server and map to the relevant handlers. Visit the 
[TODO List API Example](./examples/todo_list_app_example) to see how Cellotape 
for a more realistic use.
 
## Roadmap

- [ ] Improve the documentation with more usage details.
- [ ] Runtime validation for request body & parameters based on the OpenAPI spec.
  Considered multiple options, including:

>   - https://github.com/Go-playground/validator
>   - https://github.com/xeipuuv/Gojsonschema

Reviewer: ```suggestion
  - https://github.com/go-playground/validator
  - https://github.com/xeipuuv/gojsonschema
```
